ksqldb-common/src/test/java/io/confluent/ksql/function/UdfIndexTest.java

```diff
@@ -231,6 +238,127 @@ public void shouldChooseCorrectMap() {
     assertThat(fun.name(), equalTo(EXPECTED));
   }
 
+  @Test
+  public void shouldChooseCorrectLambdaFunction() {
+    // Given:
+    givenFunctions(
+        function(EXPECTED, false, GENERIC_LIST, LAMBDA_FUNCTION)
+    );
+
+    // When:
+    final KsqlScalarFunction fun = udfIndex.getFunction(
+        ImmutableList.of(
+            SqlArgument.of(ARRAY_ARG),
+            SqlArgument.of(
+                SqlLambda.of(
+                    ImmutableList.of(SqlTypes.STRING),
+                    INTEGER))));
+
+    // Then:
+    assertThat(fun.name(), equalTo(EXPECTED));
+  }
+
+  @Test
+  public void shouldChooseCorrectLambdaBiFunction() {
+    // Given:
+    givenFunctions(
+        function(EXPECTED, false, GENERIC_MAP, LAMBDA_BI_FUNCTION)
+    );
+
+    // When:
+    final KsqlScalarFunction fun = udfIndex.getFunction(
+        ImmutableList.of(
+            SqlArgument.of(MAP1_ARG),
+            SqlArgument.of(
+                SqlLambda.of(
+                    ImmutableList.of(SqlTypes.STRING, SqlTypes.STRING),
+                    INTEGER))));
+
+    // Then:
+    assertThat(fun.name(), equalTo(EXPECTED));
+  }
+
+  @Test
+  public void shouldChooseCorrectLambdaForTypeSpecificCollections() {
+    // Given:
+    givenFunctions(
+        function(EXPECTED, false, MAP1, LAMBDA_BI_FUNCTION_STRING)
+    );
+
+    // When:
+    final KsqlScalarFunction fun1 = udfIndex.getFunction(
+        ImmutableList.of(
+            SqlArgument.of(MAP1_ARG),
+            SqlArgument.of(
+                SqlLambda.of(
+                    ImmutableList.of(SqlTypes.STRING, SqlTypes.STRING),
+                    SqlTypes.BOOLEAN))));
+
+    final KsqlScalarFunction fun2 = udfIndex.getFunction(
+        ImmutableList.of(
+            SqlArgument.of(MAP1_ARG),
+            SqlArgument.of(
+                SqlLambda.of(
+                    ImmutableList.of(SqlTypes.STRING, SqlTypes.STRING),
+                    INTEGER))));
+
+    final Exception e = assertThrows(
+        Exception.class,
+        () -> udfIndex.getFunction(
+            ImmutableList.of(
+                SqlArgument.of(MAP1_ARG),
+                SqlArgument.of(
+                    SqlLambda.of(
+                        ImmutableList.of(SqlTypes.BOOLEAN, INTEGER),
+                        INTEGER))))
+    );
+
+    // Then:
+    assertThat(fun1.name(), equalTo(EXPECTED));
+    assertThat(fun2.name(), equalTo(EXPECTED));
+    assertThat(e.getMessage(), containsString("Valid alternatives are:"
+        + lineSeparator()
+        + "expected(MAP<VARCHAR, VARCHAR>, LAMBDA<[VARCHAR, VARCHAR], A>)"));
```

Comment:
if this is shown to users, we might want to consider `toString` on lambdas to be something like `(VARCHAR, VARCHAR) -> A` instead of `LAMBDA<[VARCHAR, VARCHAR], A>` which is somewhat difficult to read

Reply:
Do you think we should leave the `LAMBDA` identifier for the new `toString`? Just wondering if users will be able to easily identify if it's `((VARCHAR, VARCHAR) -> A`